### ISSUE TITLE
Links vs. References

### ISSUE THREAD

**Issue by SergeyKanzhelev** [bogdandrutu/openconsensus#14]
From: https://docs.google.com/document/d/1gOjNl7nYardzaQJVtJCFJUBUIOwssKN-r9IGyL7HCJc/edit#heading=h.bk2at8j2ubk8

Links and references are equivalent concepts, acting as edges to connect spans into a directed graph.
OT Reference
```
public final class References {
    public static final String CHILD_OF = "child_of";
    public static final String FOLLOWS_FROM = "follows_from";
}
SpanBuilder.addReference(String referenceType, SpanContext referencedContext);
OC Link
public abstract class Link {
  public enum Type {
    CHILD_LINKED_SPAN,
    PARENT_LINKED_SPAN
  }
  public static Link fromSpanContext(
      SpanContext context, Type type, Map<String, AttributeValue> attributes) {};
  public abstract TraceId getTraceId();
  public abstract SpanId getSpanId();
  public abstract Type getType();
  public abstract Map<String, AttributeValue> getAttributes();
}
```
Differences

OC Links can be added at any time; OT References must be added on Span creation.
Links can have Attributes; References cannot. (what is a link attribute?)
OC Link types are Child, Parent, and Unspecified; OT Reference types are ChildOf and FollowsFrom. 

Proposed Resolutions
Only add References during Span start time
In order to support Zipkin’s reuse of spanID’s for client/server span pairs, OT only allows references to be added. Because Zipkin is an important tracing system, we feel it would be unfair to not support them. Unless we can resolve zipkin support without it, this is a requirement.

Update: ZipKin no longer requires this feature.
Questions
Does OC have plans for supporting Zipkin without this restriction?
Is the OC LinkTypeParent necessary or speculative? If necessary, what is the use case?
Would the OC team be fine with this restriction?
Would the OC team be fine with adding FollowsFrom?

**Comment by SergeyKanzhelev** [comment]
[bogdan] Links are not the same as references. Links used for 
- batching - like waiting for multiple writes before doing actual IO operation. Like multi-parenting
- 

[yuri] - references are the same

[ted] follows from is for async. In OC it's separate traces. This is biggest difference

Need more discussions...

**Comment by tedsuo** [comment]
Suggestion: can we write down examples of the different scenarios we are trying to model? This will help ensure we're all on the same page, and no edge cases get left behind.

**Comment by tedsuo** [comment]
TODO: For clarity, provide code examples from OT and OC modeling each scenario.

**Comment by carlosalberto** [comment]
A small snippet for OT's `FOLLOWS_FROM` would be a `Promise` + callbacks usage:

```java
public class Promise<T> {
  public Promise() {
    // 1. Capture the active Span.
    this.parentSpan = tracer.activeSpan();
  }
  
  // 2. Add callacks that will be invoked asynchronously
  //    upon success.
  public void onSuccess(Callback<T> callback) {
    successCallbacks.add(callback);
  }
  
  public void success(final T result) {
    // 3. Upon succcess, invoke the callbacks asynchronously
    //    through a service executor. Each one will do an independent,
    //    post-processing task that is *not* core part of the parent
    //    Span/task.
    for (Callback<T> callback: successCallbacks) {
      executorService.submit(new Runnable {
        @Override
        public void run() {
          Span span = tracer.buildSpan("postProcessingTask")
            .addReference(References.FOLLOWS_FROM, parentSpan.context())
            .start();
          try (Scope scope = tracer.activateSpan(span)) {
            // 4. Do a post-processing task and create a child Span.
          } finally {
            span.finish();
          }
        }
      });
    }
  }
}
```

**Comment by SergeyKanzhelev** [comment]
Another consideration to not forget - links has a limit of 20 per span. References are unlimited

**Comment by tedsuo** [comment]
## LINKS
* Two use cases:
  * Batch processing
  * Restarting a trace
* Links are for joining multiple traces
* Links are NOT for joining spans within a trace
* Maybe do not need an Enum for this?
  * enums are more confusing that we thought initially
  * CHILD reference technically not necessary


## REFERENCES
* For span parentage only, not so much for  joining traces.
Key difference: is parent waiting or not?

Sync vs Async is better name
Boolean property of SpanBuilder `parentSpan`


Action Items:
* No more references, just maybe an `async` boolean [Carlos]
* Initial pass on new Links API [Bogdan]
    * Keep links for Trace joining batch jobs
    * Remove or rename link types
        * CHILD is confusing

**Comment by bogdandrutu** [comment]
Proposal:
Reference {  // or Link??
  SpanContext context;
  sync/async  // TBD - maybe deferred
  Attributes // 
}

SpanData {
  Reference parent;  // must have the same traceId

  List<Reference> links; // can have different or same traceId
}

SpanBuilder {
  // Keep the same traceId
  parent(SpanContext);  //
  parent(Span);  //

  // Not necessary to keep the traceId, most likely not.
  link(SpanContext);  // parentLink? linkedTrace? link?? reference??

  link(Reference);  // parentLink? linkedTrace? link?? reference??
}

**Comment by yurishkuro** [comment]
NB: sync/async is not a particularly accurate distinction. Program may be using async programming style, in which case _all spans are async_, yet parent spans are waiting on children spans, which is what's important about the child-of / follows-from distinction in OT.

I think we need a different name for this (child-of / follows-from are not particularly good  names either).

**Comment by yurishkuro** [comment]
Discussed: 

we may need three-value state of the link: 
  * parent depends on child outcome
  * parent does not depend
  * relationship is unknown

**Comment by yurishkuro** [comment]
@bogdandrutu what did you have in mind in terms of different post-processing between the last two link types? If there is no difference in processing, perhaps we don't need the third state.

**Comment by tedsuo** [comment]
@yurishkuro since this is an additive feature, perhaps we should defer this for now, and have a community discussion?

**Comment by yurishkuro** [comment]
+1

**Comment by SergeyKanzhelev** [comment]
should we close this issue in favor of https://github.com/bogdandrutu/openconsensus/issues/176?
